File: app/src/main/java/com/snakes_ladder/dice/adapters/GridAdapter.java
package com.snakes_ladder.dice.adapters;

import android.content.Context;
import android.os.CountDownTimer;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import com.snakes_ladder.dice.R;
import com.snakes_ladder.dice.models.GridInfo;
import com.snakes_ladder.dice.util.Util;

import java.util.List;

public class GridAdapter extends RecyclerView.Adapter<GridAdapter.ViewHolder> {
    private final int gridItemSize;
    private final RelativeLayout.LayoutParams layoutParams1;
    private final RelativeLayout.LayoutParams layoutParams2;
    private List<GridInfo> values;
    private int player1Position = 1;
    private int player2Position = 1;
    private boolean isPlayer1 = true;
    StatusListener statusListener;

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        public ImageView player1Icon;
        public ImageView player2Icon;

        public ViewHolder(View v) {
            super(v);
            player1Icon = v.findViewById(R.id.player1);
            player2Icon = v.findViewById(R.id.player2);
        }
    }

    // Provide a suitable constructor (depends on the kind of dataset)
    public GridAdapter(Context mContext, List<GridInfo> myDataset, int gridItemSize) {
        values = myDataset;
        this.gridItemSize = gridItemSize;

        layoutParams1 = new RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT
        );

        layoutParams2 = new RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT
        );

        try {
            this.statusListener = ((StatusListener) mContext);
            statusListener.status(isPlayer1);
        } catch (ClassCastException e) {
            throw new ClassCastException("Activity must implement StatusListener.");
        }
    }

    // Create new views (invoked by the layout manager)
    @Override
    public GridAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
                                                     int viewType) {
        // create a new view
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View iv = inflater.inflate(R.layout.row_layout, parent, false);
        LinearLayout.LayoutParams parms = new LinearLayout.LayoutParams(gridItemSize, gridItemSize);
        iv.setLayoutParams(parms);

        // set the view's size, margins, paddings and layout parameters
        return new ViewHolder(iv);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(ViewHolder holder, final int position) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        final GridInfo gridInfo = values.get(position);

        if (player1Position == player2Position) {
            layoutParams1.setMargins(0, 0, Util.convertDpToPixel(30), Util.convertDpToPixel(10));
            holder.player1Icon.setLayoutParams(layoutParams1);
            layoutParams2.setMargins(Util.convertDpToPixel(30), Util.convertDpToPixel(10), 0, 0);
            holder.player2Icon.setLayoutParams(layoutParams2);
        } else {
            layoutParams1.setMargins(0, 0, 0, 0);
            holder.player1Icon.setLayoutParams(layoutParams1);
            holder.player2Icon.setLayoutParams(layoutParams1);
        }

        if (gridInfo.getPosition() != player1Position)
            holder.player1Icon.setImageDrawable(null);
        else
            holder.player1Icon.setImageResource(R.drawable.ic_strategy_1);

        if (gridInfo.getPosition() != player2Position)
            holder.player2Icon.setImageDrawable(null);
        else
            holder.player2Icon.setImageResource(R.drawable.ic_strategy_2);
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return values.size();
    }

    public void setPlayerPosition(int newPosition, int moveCount, ImageView dice_picture) {

        if (isPlayer1) {

            if (player1Position == 1 && moveCount == 1)
                movePlayer(newPosition, moveCount, dice_picture);
            else if (player1Position > 1) {
                movePlayer(newPosition, moveCount, dice_picture);
            } else {
                isPlayer1 = false;
                statusListener.status(false);
            }

        } else {

            if (player2Position == 1 && moveCount == 1)
                movePlayer(newPosition, moveCount, dice_picture);
            else if (player2Position > 1) {
                movePlayer(newPosition, moveCount, dice_picture);
            } else {
                isPlayer1 = true;
                statusListener.status(true);
            }
        }
    }

    private void movePlayer(final int newPosition, int moveCount, final ImageView dice_picture) {
        dice_picture.setEnabled(false);
        new CountDownTimer((moveCount * 500) + 500, 500) {
            public void onTick(long millisUntilFinished) {
                if (isPlayer1)
                    player1Position += 1;
                else
                    player2Position += 1;
                notifyDataSetChanged();
            }

            public void onFinish() {
                if (isPlayer1)
                    player1Position = newPosition;
                else player2Position = newPosition;

                notifyDataSetChanged();

                if (newPosition == 25)
                    statusListener.winner(isPlayer1);
                else
                    isPlayer1 = !isPlayer1;

                statusListener.status(isPlayer1);
                dice_picture.setEnabled(true);
            }
        }.start();
    }

    public int getPlayerPosition() {
        if (isPlayer1)
            return player1Position;
        else return player2Position;
    }

    public interface StatusListener {
        void winner(boolean isPlayer1);
        void loser(boolean isPlayer1);
        void status(boolean isPlayer1);
    }
}

File: app/src/main/java/com/snakes_ladder/dice/adapters/GridAdapter.kt
package com.snakes_ladder.dice.adapters

import android.content.Context
import android.os.CountDownTimer
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout

import com.snakes_ladder.dice.R
import com.snakes_ladder.dice.models.GridInfo
import com.snakes_ladder.dice.util.Util

class GridAdapter// Provide a suitable constructor (depends on the kind of dataset)
(mContext: Context?, private val values: List<GridInfo>, private val gridItemSize: Int) : RecyclerView.Adapter<GridAdapter.ViewHolder>() {
    private val layoutParams1: RelativeLayout.LayoutParams
    private val layoutParams2: RelativeLayout.LayoutParams
    private var player1Position = 1
    private var player2Position = 1
    private var isPlayer1 = true
    internal var statusListener: StatusListener

    val playerPosition: Int
        get() = if (isPlayer1)
            player1Position
        else
            player2Position

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    inner class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        // each data item is just a string in this case
        var player1Icon: ImageView = v.findViewById(R.id.player1)
        var player2Icon: ImageView = v.findViewById(R.id.player2)

    }

    init {

        layoutParams1 = RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT
        )

        layoutParams2 = RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.MATCH_PARENT,
                RelativeLayout.LayoutParams.MATCH_PARENT
        )

        try {
            this.statusListener = mContext as StatusListener
            statusListener.status(isPlayer1)
        } catch (e: ClassCastException) {
            throw ClassCastException("Activity must implement StatusListener.")
        }

    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): GridAdapter.ViewHolder {
        // create a new view
        val inflater = LayoutInflater.from(parent.context)
        val iv = inflater.inflate(R.layout.row_layout, parent, false)
        val parms = LinearLayout.LayoutParams(gridItemSize, gridItemSize)
        iv.layoutParams = parms

        // set the view's size, margins, paddings and layout parameters
        return ViewHolder(iv)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        val gridInfo = values[position]

        if (player1Position == player2Position) {
            layoutParams1.setMargins(0, 0, Util.convertDpToPixel(30f), Util.convertDpToPixel(10f))
            holder.player1Icon.layoutParams = layoutParams1
            layoutParams2.setMargins(Util.convertDpToPixel(30f), Util.convertDpToPixel(10f), 0, 0)
            holder.player2Icon.layoutParams = layoutParams2
        } else {
            layoutParams1.setMargins(0, 0, 0, 0)
            holder.player1Icon.layoutParams = layoutParams1
            holder.player2Icon.layoutParams = layoutParams1
        }

        if (gridInfo.position != player1Position)
            holder.player1Icon.setImageDrawable(null)
        else
            holder.player1Icon.setImageResource(R.drawable.ic_strategy_1)

        if (gridInfo.position != player2Position)
            holder.player2Icon.setImageDrawable(null)
        else
            holder.player2Icon.setImageResource(R.drawable.ic_strategy_2)
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return values.size
    }

    fun setPlayerPosition(newPosition: Int, moveCount: Int, dice_picture: ImageView) {

        if (isPlayer1) {

            if (player1Position == 1 && moveCount == 1)
                movePlayer(newPosition, moveCount, dice_picture)
            else if (player1Position > 1) {
                movePlayer(newPosition, moveCount, dice_picture)
            } else {
                isPlayer1 = false
                statusListener.status(false)
            }

        } else {

            if (player2Position == 1 && moveCount == 1)
                movePlayer(newPosition, moveCount, dice_picture)
            else if (player2Position > 1) {
                movePlayer(newPosition, moveCount, dice_picture)
            } else {
                isPlayer1 = true
                statusListener.status(true)
            }
        }
    }

    private fun movePlayer(newPosition: Int, moveCount: Int, dice_picture: ImageView) {
        dice_picture.isEnabled = false
        object : CountDownTimer(((moveCount * 400) + 400).toLong(), 400) {
            override fun onTick(millisUntilFinished: Long) {
                if (isPlayer1) {
                    if (newPosition > player1Position)
                        player1Position += 1
                } else {
                    if (newPosition > player2Position)
                        player2Position += 1
                }
                notifyDataSetChanged()
            }

            override fun onFinish() {
                if (isPlayer1)
                    player1Position = newPosition
                else
                    player2Position = newPosition

                notifyDataSetChanged()

                if (newPosition == 25)
                    statusListener.winner(isPlayer1)
                else
                    isPlayer1 = !isPlayer1

                statusListener.status(isPlayer1)
                dice_picture.isEnabled = true
            }
        }.start()
    }

    interface StatusListener {
        fun winner(isPlayer1: Boolean)
        fun loser(isPlayer1: Boolean)
        fun status(isPlayer1: Boolean)
    }
}